src/interpreter/internal.rs

```diff
@@ -166,10 +166,28 @@ impl<S> Interpreter<S> {
         }?;
 
         // update the local copy of the output
-        let mut output = Output::variable(owner_to_set, amount_to_set, asset_id_to_update);
+        let output = Output::variable(owner_to_set, amount_to_set, asset_id_to_update);
+        self.set_output(out_idx, output)
+    }
+
+    /// update an output both in serialized form in-memory and on the referenced tx
+    pub(crate) fn set_output(&mut self, out_idx: usize, mut output: Output) -> Result<(), RuntimeError> {
+        // verify output type matches out_idx type
+        // Use halt errors as these should already have been validated
+        let tx_out = self.tx.outputs().get(out_idx).ok_or(RuntimeError::Halt(io::Error::new(
+            ErrorKind::Other,
+            "Invalid output index",
+        )))?;
+        if std::mem::discriminant(tx_out) != std::mem::discriminant(&output) {
+            // TODO: need a panic reason for a generic UnexpectedOutputType
```

Comment:
Could use `PanicReason::ExpectedOutputVariable`?

Reply:
oops that todo isn't needed anymore, since these are halting errors that should've been prevented earlier.